DR-007: mTLS for Distributed Query Cluster Communication (#8722)
spiceai/spiceai: docs/decisions/007-cluster-mtls.md

## Design

### Port Separation

| Port  | Visibility | Services                                            | mTLS Required              |
| ----- | ---------- | --------------------------------------------------- | -------------------------- |
| 50051 | Public     | `FlightServiceServer` (user queries), `OtelService` | Optional (user TLS config) |
| 8090  | Public     | HTTP API (REST queries, health, status)             | Optional (user TLS config) |
| 9090  | Public     | Prometheus metrics                                  | No                         |
| 50052 | Internal   | `SchedulerGrpcServer`, `ClusterService`             | **Required**               |

### CLI Arguments

```bash
# Role selection
--role {scheduler|executor}               # Explicit role (executor implied if --scheduler-address set)

# mTLS configuration (required unless --allow-insecure-connections)
--node-mtls-ca-certificate-file <path>    # CA cert to validate peer certificates
--node-mtls-certificate-file <path>       # This node's certificate
--node-mtls-key-file <path>               # This node's private key

# Network configuration

> --node-advertise-address <host:port>      # How other nodes reach this node

Reviewer: ```suggestion
--node-advertise-address <host>      # How other nodes reach this node
```

Part of my simplification effort is to remove the port here - as a reference, the Kubernetes API server doesn't allow configuring a separate port for the advertise address and just uses it from the bind address.

--node-bind-address <bind_address>        # Internal service bind (default: 0.0.0.0:50052)

# Executor-specific

> --scheduler-address <url>                 # Scheduler's internal cluster URL

Reviewer: ```suggestion
--scheduler-address <url>                 # Scheduler's internal load-balanced/virtual IP cluster URL
```

| `--cluster-scheduler-url`       | `--scheduler-address`             |

### New CLI Behavior

* `--scheduler-address`: If set, `--role executor` is implied and can be omitted.
* `--allow-insecure-connections`: Disables mTLS requirement for dev/test (WARNING: never use in production).

### Removed CLI Arguments

* `--cluster-api-key`: Replaced by mTLS.

### New gRPC Service

* `ClusterService` with `GetAppDefinition` and `ExpandSecret` RPCs.
* Served on port 50052 with mTLS.
* `SchedulerGrpcServer` (Ballista) also moved to port 50052.

### Flight Actions Removed

* `GetAppDefinition` Flight Action removed from public port.
* `ExpandSecret` Flight Action removed from public port.
* These are security-sensitive and should never be on the public port.

### TLS Configuration Structure

```rust
pub struct ClusterTlsConfig {
    /// CA certificate used to validate other cluster nodes
    pub ca_certificate: Certificate,
    /// Client TLS config with CA and client identity for mTLS
    pub client_tls_config: ClientTlsConfig,
    /// Server identity (cert + key) for serving TLS
    pub server_identity: Identity,
}
```

### Server Configuration

> * Scheduler starts internal cluster server on `--cluster-address` (default 50052).

Reviewer: ```suggestion
* Scheduler starts internal cluster server on `--node-bind-address` (default 0.0.0.0:50052).
```

> * Executor starts flight server on `--cluster-address` with mTLS.

Reviewer: ```suggestion
* Executor starts flight server on `--node-bind-address` with mTLS.
```
